aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp

```diff
@@ -0,0 +1,122 @@
+#include <ATen/ATen.h>
+#include <ATen/NamedTensorUtils.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/native/nestedtensor/NestedTensorImpl.h>
+#include <c10/core/DispatchKey.h>
+
+namespace at {
+namespace native {
+
+NestedTensorImpl::NestedTensorImpl(
+    at::Tensor buffer,
+    at::Tensor nested_size_tensor)
+    : TensorImpl(
+          c10::DispatchKeySet({DispatchKey::NestedTensor}),
+          buffer.dtype(),
+          buffer.device()),
+      _buffer(buffer),
+      _nested_size_tensor(nested_size_tensor) {
+  remove_autograd_key();
+  key_set_ =
+      key_set_ - c10::DispatchKeySet({c10::DispatchKey::ADInplaceOrView});
+}
+
+at::Tensor wrap_buffer(at::Tensor buffer, at::Tensor nested_size_tensor) {
+  TORCH_CHECK(buffer.is_contiguous(), "Given buffer must be contiguous.");
+  return at::detail::make_tensor<NestedTensorImpl>(
+      std::move(buffer), nested_size_tensor);
+}
+
+std::vector<at::Tensor> NestedTensor_unbind(
+    const at::Tensor& self,
+    int64_t dim) {
+  TORCH_CHECK(
+      dim == 0,
+      "NestedTensor can only be unbound along dimension 0 ",
+      "got dimension ",
+      dim,
+      " instead.");
+  auto esizes = get_nested_size_tensor(self);
+  auto buffer = get_buffer(self);
+  std::vector<at::Tensor> result_tensors;
+  if (esizes.dim() == 0) {
+    return result_tensors;
+  }
+  auto esizes_chunks = esizes.unbind(0);
+  std::vector<int64_t> splits;
+  for (int64_t i = 0; i < esizes_chunks.size(); i++) {
+    splits.push_back(esizes_chunks[i].prod().item<int64_t>());
+  }
+  auto buffer_chunks = at::split_with_sizes(buffer, IntArrayRef(splits));
+  for (int64_t i = 0; i < buffer_chunks.size(); i++) {
+    auto esize_chunk = esizes_chunks[i];
+    std::vector<int64_t> esize_vector(
+        esize_chunk.data_ptr<int64_t>(),
+        esize_chunk.data_ptr<int64_t>() + esize_chunk.numel());
+    result_tensors.push_back(buffer_chunks[i].view(IntArrayRef(esize_vector)));
```

Comment:
no need for the vector; make an IntArrayRef directly

Reply:
IntArrayRef can't be constructed from a Tensor.

```
    ../aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp: In function ‘std::vector<at::Tensor> at::native::NestedTensor_unbind(const at::Tensor&, int64_t)
’:
    ../aten/src/ATen/native/nestedtensor/NestedTensorImpl.cpp:54:75: error: no matching function for call to ‘c10::ArrayRef<long int>::ArrayRef(const at::Tenso
r&)’
         result_tensors.push_back(buffer_chunks[i].view(IntArrayRef(esize_chunk)));                                                                          ^
```

Reply:
you can construct one using the `ArrayRef(const T* begin, const T* end)` constructor just as you are constructing the vector here:

```
  IntArrayRef esize_arrayref(
        esize_chunk.data_ptr<int64_t>(),
        esize_chunk.data_ptr<int64_t>() + esize_chunk.numel());
```